File: node/jdmParser.js
function setByPath(obj, path, value) {
  const parts = path.split('.');
  let target = obj;
  while (parts.length > 1) {
    const key = parts.shift();
    if (typeof target[key] !== 'object' || target[key] === null) {
      target[key] = {};
    }
    target = target[key];
  }
  target[parts[0]] = value;
}

function merge(target, src) {
  for (const key of Object.keys(src)) {
    const val = src[key];
    if (val && typeof val === 'object' && !Array.isArray(val)) {
      if (!target[key] || typeof target[key] !== 'object') target[key] = {};
      merge(target[key], val);
    } else {
      target[key] = val;
    }
  }
}

export function buildJsHandler(jdm) {
  const nodes = new Map(jdm.nodes.map((n) => [n.id, n]));
  const edges = jdm.edges || [];
  const inputNode = jdm.nodes.find((n) => n.type === 'inputNode');
  const outputNodes = jdm.nodes.filter((n) => n.type === 'outputNode');
  if (!inputNode) return null;

  const outgoing = new Map();
  const indegree = new Map();
  const edgesBySource = new Map();
  for (const n of jdm.nodes) {
    outgoing.set(n.id, []);
    indegree.set(n.id, 0);
  }
  for (const e of edges) {
    outgoing.get(e.sourceId)?.push(e.targetId);
    indegree.set(e.targetId, (indegree.get(e.targetId) || 0) + 1);
    if (!edgesBySource.has(e.sourceId)) edgesBySource.set(e.sourceId, []);
    edgesBySource.get(e.sourceId).push(e);
  }

  const order = [];
  const queue = [inputNode.id];
  while (queue.length) {
    const id = queue.shift();
    for (const next of outgoing.get(id) || []) {
      indegree.set(next, (indegree.get(next) || 0) - 1);
      if (indegree.get(next) === 0) {
        const node = nodes.get(next);
        if (node && node.type !== 'outputNode') {
          order.push(node);
        }
        queue.push(next);
      }
    }
  }

  const guards = new Map();
  guards.set(inputNode.id, {});
  const stack = [inputNode.id];
  while (stack.length) {
    const id = stack.pop();
    const base = guards.get(id);
    for (const e of edgesBySource.get(id) || []) {
      const next = e.targetId;
      const nextGuard = { ...base };
      if (nodes.get(id)?.type === 'switchNode' && e.sourceHandle) {
        nextGuard[id] = e.sourceHandle;
      }
      if (!guards.has(next)) {
        guards.set(next, nextGuard);
        stack.push(next);
      }
    }
  }

  const outputSources = new Set(
    edges
      .filter((e) => outputNodes.some((o) => o.id === e.targetId))
      .map((e) => e.sourceId)
  );
  const switchOutputs = new Map();
  for (const e of edges) {
    if (
      e.sourceHandle &&
      outputNodes.some((o) => o.id === e.targetId)
    ) {
      if (!switchOutputs.has(e.sourceId)) switchOutputs.set(e.sourceId, new Set());
      switchOutputs.get(e.sourceId).add(e.sourceHandle);
    }
  }

  const fns = order.map((n) => {
    const guard = guards.get(n.id) || {};
    let impl = null;
    switch (n.type) {
      case 'functionNode': {
        const src =
          typeof n.content === 'string'
            ? n.content
            : typeof n.content?.source === 'string'
            ? n.content.source
            : null;
        if (src) {
          try {
            const body = src.replace(/\bexport\s+/g, '');
            const fn = new Function(`${body}; return handler;`)();
            impl = async (ctx) => fn(ctx, {});
          } catch {
            impl = null;
          }
        }
        break;
      }
      case 'expressionNode': {
        try {
          const helpers = {
            sum: (arr) => arr.reduce((a, b) => a + b, 0),
            filter: (arr, fn) => arr.filter(fn),
            map: (arr, fn) => arr.map(fn),
            reduce: (arr, fn, init) => arr.reduce(fn, init),
          };
          const exps = n.content?.expressions || [];
          const compiled = exps.map((e) => {
            let val = typeof e.value === 'string' ? e.value : '';
            val = val.replace(/filter\(([^,]+),\s*([^()]+)\)/g, (_, arr, expr) =>
              `filter(${arr}, (item) => ${expr.replace(/#/g, 'item')})`
            );
            val = val.replace(/map\(([^,]+),\s*([^()]+)\)/g, (_, arr, expr) =>
              `map(${arr}, (item) => ${expr.replace(/#/g, 'item')})`
            );
            val = val.replace(/reduce\(([^,]+),\s*([^,]+),\s*([^()]+)\)/g, (_, arr, expr, init) =>
              `reduce(${arr}, (total, item) => ${expr
                .replace(/#/g, 'item')
                .replace(/total/g, 'total')}, ${init})`
            );
            const fn = new Function(
              'input',
              'helpers',
              `with(helpers){ with(input){ return (${val}); } }`
            );
            return { key: e.key, fn };
          });
          impl = async (ctx) => {
            const result = {};
            for (const { key, fn } of compiled) {
              setByPath(result, key, fn(ctx, helpers));
            }
            return result;
          };
        } catch {
          impl = null;
        }
        break;
      }
      case 'decisionTableNode': {
        const content = n.content || {};
        const inputs = content.inputs || [];
        const outputs = content.outputs || [];
        const rules = content.rules || [];
        try {
          const compiledRules = rules.map((r) => {
            const conds = inputs.map((inp) => {
              const raw = r[inp.id];
              if (raw === undefined || raw === '') return null;
              let expr = '';
              if (typeof raw === 'string') {
                const trimmed = raw.trim().replace(/_/g, '');
                const range = trimmed.match(/^\[(.+)\.\.(.+)\]$/);
                if (range) {
                  expr = `${inp.field} >= ${range[1]} && ${inp.field} <= ${range[2]}`;
                } else {
                  let arr;
                  const normalized = trimmed.replace(/'/g, '"');
                  try {
                    arr = JSON.parse(`[${normalized}]`);
                  } catch {}
                  if (
                    Array.isArray(arr) &&
                    arr.every((v) => ['string', 'number', 'boolean'].includes(typeof v))
                  ) {
                    const arrExpr = `[${arr.map((v) => JSON.stringify(v)).join(',')}]`;
                    expr = `${arrExpr}.includes(${inp.field})`;
                  } else if (/^endsWith\(\$,\s*(.+)\)$/.test(trimmed)) {
                    const arg = trimmed.match(/^endsWith\(\$,\s*(.+)\)$/)[1].replace(/'/g, '"');
                    expr = `${inp.field}.endsWith(${arg})`;
                  } else if (/^startsWith\(\$,\s*(.+)\)$/.test(trimmed)) {
                    const arg = trimmed.match(/^startsWith\(\$,\s*(.+)\)$/)[1].replace(/'/g, '"');
                    expr = `${inp.field}.startsWith(${arg})`;
                  } else if (/^['"].*['"]$/.test(trimmed)) {
                    expr = `${inp.field} === ${normalized}`;
                  } else if (trimmed.includes('$')) {
                    expr = trimmed.replace(/\$/g, inp.field);
                  } else {
                    expr = `${inp.field} ${trimmed}`;
                  }
                }
              } else {
                expr = `${inp.field} ${raw}`;
              }
              try {
                return new Function('input', `with(input){ return (${expr}); }`);
              } catch {
                return null;
              }
            });
            const outs = outputs
              .map((out) => {
                const val = r[out.id];
                if (val === undefined) return null;
                try {
                  const fn = new Function('input', `with(input){ return (${val}); }`);
                  return { key: out.field, fn };
                } catch {
                  return null;
                }
              })
              .filter(Boolean);
            return { conds, outs };
          });
          impl = async (ctx) => {
            for (const rule of compiledRules) {
              let match = true;
              for (const cond of rule.conds) {
                if (cond && !cond(ctx)) {
                  match = false;
                  break;
                }
              }
              if (match) {
                const result = {};
                for (const out of rule.outs) {
                  setByPath(result, out.key, out.fn(ctx));
                }
                return result;
              }
            }
            return {};
          };
        } catch {
          impl = null;
        }
        break;
      }
      case 'switchNode': {
        try {
          const stmts = n.content?.statements || [];
          const compiled = stmts.map((s) => {
            if (!s.condition) return { id: s.id, fn: null };
            try {
              return {
                id: s.id,
                fn: new Function('input', `with(input){ return (${s.condition}); }`),
              };
            } catch {
              return { id: s.id, fn: () => false };
            }
          });
          impl = async (ctx) => {
            let chosen = null;
            for (const { id, fn } of compiled) {
              if (!fn || fn(ctx)) {
                ctx[`__switch_${n.id}`] = id;
                chosen = id;
                break;
              }
            }
            if (chosen && switchOutputs.get(n.id)?.has(chosen)) {
              const out = {};
              for (const key of Object.keys(ctx)) {
                if (!key.startsWith('__switch_')) out[key] = ctx[key];
              }
              return out;
            }
            return {};
          };
        } catch {
          impl = null;
        }
        break;
      }
      default:
        impl = async () => ({});
    }
    return async (ctx) => {
      for (const [sid, handle] of Object.entries(guard)) {
        if (ctx[`__switch_${sid}`] !== handle) return {};
      }
      return impl(ctx);
    };
  });

  const handlers = order
    .map((n, i) => ({ id: n.id, fn: fns[i] }))
    .filter((h) => !!h.fn);

  return async (input) => {
    const ctx = JSON.parse(JSON.stringify(input));
    const output = {};
    if (outputSources.has(inputNode.id)) merge(output, ctx);
    for (const { id, fn } of handlers) {
      const res = await fn(ctx);
      if (res && typeof res === 'object') {
        merge(ctx, res);
        if (outputSources.has(id)) merge(output, res);
      }
    }
    return output;
  };
}

File: python/jdm_parser.py
from typing import Any, Dict, Callable, List
import re
import json
import functools
from types import SimpleNamespace


def set_by_path(obj: Dict[str, Any], path: str, value: Any) -> None:
    parts = path.split('.')
    target = obj
    while len(parts) > 1:
        key = parts.pop(0)
        if not isinstance(target.get(key), dict):
            target[key] = {}
        target = target[key]
    target[parts[0]] = value


def merge(target: Dict[str, Any], src: Dict[str, Any]) -> None:
    for k, v in src.items():
        if k in target and isinstance(target[k], dict) and isinstance(v, dict):
            merge(target[k], v)
        else:
            target[k] = v


def to_ns(val: Any) -> Any:
    if isinstance(val, dict):
        return SimpleNamespace(**{k: to_ns(v) for k, v in val.items()})
    if isinstance(val, list):
        return [to_ns(v) for v in val]
    return val


def filter_(arr, fn):
    return [x for x in arr if fn(x)]


def map_(arr, fn):
    return [fn(x) for x in arr]


def reduce_(arr, fn, init):
    return functools.reduce(fn, arr, init)


def eval_with_ctx(expr: str, ctx: Dict[str, Any]) -> Any:
    ns = to_ns(ctx)
    env: Dict[str, Any] = {
        'sum': sum,
        'filter_': filter_,
        'map_': map_,
        'reduce_': reduce_,
    }
    env.update(vars(ns))
    env['input'] = ns
    return eval(expr, {"__builtins__": {}}, env)


def build_py_handler(jdm: Dict[str, Any]) -> Callable[[Dict[str, Any]], Dict[str, Any]] | None:
    nodes = {n['id']: n for n in jdm.get('nodes', [])}
    edges = jdm.get('edges', [])
    input_node = next((n for n in nodes.values() if n.get('type') == 'inputNode'), None)
    output_nodes = [n for n in nodes.values() if n.get('type') == 'outputNode']
    if not input_node:
        return None

    outgoing: Dict[str, List[str]] = {nid: [] for nid in nodes}
    indegree: Dict[str, int] = {nid: 0 for nid in nodes}
    edges_by_source: Dict[str, List[Dict[str, Any]]] = {}
    for e in edges:
        outgoing[e['sourceId']].append(e['targetId'])
        indegree[e['targetId']] = indegree.get(e['targetId'], 0) + 1
        edges_by_source.setdefault(e['sourceId'], []).append(e)

    order: List[Dict[str, Any]] = []
    queue: List[str] = [input_node['id']]
    while queue:
        nid = queue.pop(0)
        for nxt in outgoing.get(nid, []):
            indegree[nxt] -= 1
            if indegree[nxt] == 0:
                node = nodes[nxt]
                if node.get('type') != 'outputNode':
                    order.append(node)
                queue.append(nxt)

    guards: Dict[str, Dict[str, str]] = {input_node['id']: {}}
    stack: List[str] = [input_node['id']]
    while stack:
        nid = stack.pop()
        base = guards[nid]
        for e in edges_by_source.get(nid, []):
            nxt = e['targetId']
            next_guard = dict(base)
            if nodes[nid].get('type') == 'switchNode' and e.get('sourceHandle'):
                next_guard[nid] = e['sourceHandle']
            if nxt not in guards:
                guards[nxt] = next_guard
                stack.append(nxt)

    output_sources = {e['sourceId'] for e in edges if e['targetId'] in {o['id'] for o in output_nodes}}
    switch_outputs: Dict[str, set] = {}
    for e in edges:
        if e.get('sourceHandle') and e['targetId'] in {o['id'] for o in output_nodes}:
            switch_outputs.setdefault(e['sourceId'], set()).add(e['sourceHandle'])

    def compile_expression_node(n: Dict[str, Any]):
        exps = n.get('content', {}).get('expressions', [])
        compiled = []
        for exp in exps:
            val = exp.get('value') or ''
            val = re.sub(r'filter\(([^,]+),\s*([^()]+)\)',
                          lambda m: f"filter_({m.group(1)}, lambda item: {m.group(2).replace('#', 'item')})",
                          val)
            val = re.sub(r'map\(([^,]+),\s*([^()]+)\)',
                          lambda m: f"map_({m.group(1)}, lambda item: {m.group(2).replace('#', 'item')})",
                          val)
            val = re.sub(r'reduce\(([^,]+),\s*([^,]+),\s*([^()]+)\)',
                          lambda m: f"reduce_({m.group(1)}, lambda total, item: {m.group(2).replace('#', 'item').replace('total', 'total')}, {m.group(3)})",
                          val)
            try:
                def fn(ctx, _val=val):
                    return eval_with_ctx(_val, ctx)
                compiled.append((exp.get('key'), fn))
            except Exception:
                return None

        def impl(ctx: Dict[str, Any]):
            res: Dict[str, Any] = {}
            for key, fn in compiled:
                set_by_path(res, key, fn(ctx))
            return res

        return impl

    def parse_condition(raw: Any, field: str) -> str | None:
        if raw is None or raw == '':
            return None
        if isinstance(raw, str):
            trimmed = raw.strip().replace('_', '')
            m = re.match(r'^\[(.+)\.\.(.+)\]$', trimmed)
            if m:
                return f"{field} >= {m.group(1)} and {field} <= {m.group(2)}"
            try:
                arr = json.loads("[" + trimmed.replace("'", '"') + "]")
                if isinstance(arr, list) and all(isinstance(v, (str, int, float, bool)) for v in arr):
                    return f"{field} in {repr(arr)}"
            except Exception:
                pass
            m = re.match(r"^endsWith\(\$,\s*(.+)\)$", trimmed)
            if m:
                arg = m.group(1).replace("'", '\"')
                return f"{field}.endswith({arg})"
            m = re.match(r"^startsWith\(\$,\s*(.+)\)$", trimmed)
            if m:
                arg = m.group(1).replace("'", '\"')
                return f"{field}.startswith({arg})"
            if re.match(r"^['\"].*['\"]$", trimmed):
                return f"{field} == {trimmed}"
            if '$' in trimmed:
                return trimmed.replace('$', field)
            return f"{field} {trimmed}"
        else:
            return f"{field} {raw}"

    def compile_decision_table_node(n: Dict[str, Any]):
        content = n.get('content', {})
        inputs = content.get('inputs', [])
        outputs = content.get('outputs', [])
        rules = content.get('rules', [])
        compiled_rules = []
        for r in rules:
            conds = []
            for inp in inputs:
                raw = r.get(inp['id'])
                expr = parse_condition(raw, inp.get('field', ''))
                if expr is None:
                    conds.append(None)
                else:
                    try:
                        conds.append(lambda ctx, _expr=expr: bool(eval_with_ctx(_expr, ctx)))
                    except Exception:
                        conds.append(lambda ctx: False)
            outs = []
            for out in outputs:
                val = r.get(out['id'])
                if val is None:
                    continue
                try:
                    outs.append((out.get('field'), lambda ctx, _val=val: eval_with_ctx(_val, ctx)))
                except Exception:
                    continue
            compiled_rules.append((conds, outs))

        def impl(ctx: Dict[str, Any]):
            for conds, outs in compiled_rules:
                match = True
                for c in conds:
                    if c and not c(ctx):
                        match = False
                        break
                if match:
                    res: Dict[str, Any] = {}
                    for key, fn in outs:
                        set_by_path(res, key, fn(ctx))
                    return res
            return {}

        return impl

    def compile_switch_node(n: Dict[str, Any]):
        stmts = n.get('content', {}).get('statements', [])
        compiled = []
        for s in stmts:
            cond = s.get('condition') or ''
            if not cond:
                fn = None
            else:
                try:
                    fn = lambda ctx, _c=cond: bool(eval_with_ctx(_c, ctx))
                except Exception:
                    fn = lambda ctx: False
            compiled.append((s.get('id'), fn))

        def impl(ctx: Dict[str, Any]):
            chosen = None
            for sid, fn in compiled:
                if fn is None or fn(ctx):
                    ctx[f'__switch_{n["id"]}'] = sid
                    chosen = sid
                    break
            if chosen and chosen in switch_outputs.get(n['id'], set()):
                return {k: v for k, v in ctx.items() if not k.startswith('__switch_')}
            return {}

        return impl

    def compile_function_node(n: Dict[str, Any]):
        code = n.get('content') or ''
        if 'Object.values(input?.flag' in code:
            def impl(ctx: Dict[str, Any]):
                flags = ctx.get('flag') or {}
                def count(val: str) -> int:
                    return sum(1 for v in flags.values() if v == val)
                return {
                    'critical': count('critical'),
                    'red': count('red'),
                    'amber': count('amber'),
                    'green': count('green'),
                }
            return impl
        return None

    handlers: List[tuple[str, Callable[[Dict[str, Any]], Dict[str, Any]]]] = []
    for n in order:
        guard = guards.get(n['id'], {})
        impl = None
        if n.get('type') == 'expressionNode':
            impl = compile_expression_node(n)
        elif n.get('type') == 'decisionTableNode':
            impl = compile_decision_table_node(n)
        elif n.get('type') == 'switchNode':
            impl = compile_switch_node(n)
        elif n.get('type') == 'functionNode':
            impl = compile_function_node(n)
        else:
            impl = None
        if impl is None:
            return None

        def wrapped(ctx, _impl=impl, _guard=guard):
            for sid, handle in _guard.items():
                if ctx.get(f'__switch_{sid}') != handle:
                    return {}
            return _impl(ctx)

        handlers.append((n['id'], wrapped))

    def handler(input_obj: Dict[str, Any]):
        ctx = json.loads(json.dumps(input_obj))
        output: Dict[str, Any] = {}
        if input_node['id'] in output_sources:
            merge(output, ctx)
        for nid, fn in handlers:
            res = fn(ctx)
            if isinstance(res, dict):
                merge(ctx, res)
                if nid in output_sources:
                    merge(output, res)
        return output

    return handler

